### PR TITLE
Issue 24738 push files

### LINKED ISSUE #24738: dotCMS CLI: Implement endpoint to push files to dotCMS
### Parent Issue

#24753

### User Story

As a user, I'd like to have an endpoint to push files from my local environment to dotCMS using the CLI. This endpoint should: 

- Create any nested folders needed in dotCMS
- Receive an optional desiredState parameter with values: live or working (live by default if not specified)
- Also, it should receive an optional language (defaults to default language, if not specified)
- If a file is archived, we should unarchive/publish it

### Acceptance Criteria

The endpoint should support these operations: 

- [ ] Create any nested folders needed in dotCMS
- [ ] Receive an optional desiredState parameter with values: live or working (live by default if not specified) and save the files considering this status
- [ ] Also, it should receive an optional language (defaults to default language, if not specified)
- [ ] If a file is archived, we should unarchive/publish it
- [ ] Make sure zero length files can be pushed. Related to #24042

Besides, tests should be implemented accordingly.

### Proposed Objective

Core Features

### Proposed Priority

Priority 3 - Average

### PR DESCRIPTION
**Pull request by fabrizzio-dotCMS**
### Proposed Changes
* New Resource that allows pulling info about our resources using site name and folder path structure 
* We can now push, pull, and delete files and or their versions. Specifying a file asset path 
* We try to replicate as much as possible the behavior presented by our Webdav endpoints

### <samp>🤖 Generated by Copilot at a8399fb</samp>

The pull request adds new features and tests to the browser API and the web asset resource. It allows filtering folder content by file name, language, and status, and managing assets via the REST API. It also refactors some classes and adds javadoc comments and annotations. The main files affected are `BrowserAPIImpl.java`, `WebAssetResource.java`, `AssetPathResolver.java`, and their corresponding test classes.

### <samp>🤖 Generated by Copilot at a8399fb</samp>

*  Add new endpoints for downloading, uploading, and deleting assets in the WebAssetResource class ([link](https://github.com/dotCMS/core/pull/25164/files?diff=unified&w=0#diff-0d33ff04345dc5e8b42ccf0c2e516b3a20c068a52254564c8e21ee8fcbeb9d65L52-R141)), using the WebAssetHelper class and the data classes from the asset package.

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/asset/AbstractFileUploadParamsJson.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package com.dotcms.rest.api.v1.asset;
+
+
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import org.immutables.value.Value;
+
+@Value.Style(typeImmutable="*", typeAbstract="Abstract*")
+@Value.Immutable
+@JsonDeserialize(as = FileUploadParamsJson.Builder.class)
+public interface AbstractFileUploadParamsJson {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/asset/AbstractFileUploadParamsJson.java:10]: Doc me

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/asset/FileUploadData.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
+package com.dotcms.rest.api.v1.asset;
+
+import java.io.InputStream;
+import org.glassfish.jersey.media.multipart.FormDataContentDisposition;
+import org.glassfish.jersey.media.multipart.FormDataParam;
+
+public class FileUploadData {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/asset/FileUploadData.java:7]: Doc me

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/asset/WebAssetHelper.java [hunk 3]
```diff
@@ -234,6 +250,94 @@ FolderView toAssetsFolder(final Folder folder) {
...
+        final FileUploadDetail detail = form.getDetail();
+        final String assetPath = detail.getAssetPath();
+        final ResolvedAssetAndPath assetAndPath = AssetPathResolver.newInstance()
+                .resolve(assetPath, user, true);
+
+        final Host host = assetAndPath.resolvedHost();
+        final Folder folder = assetAndPath.resolvedFolder();
+        final String assetName =  fileName != null ? fileName : assetAndPath.asset();
+
+
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/asset/WebAssetHelper.java:268]: Not sure if the update a file should follows the same rules as temp api does

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/asset/WebAssetResource.java [hunk 4]
```diff
@@ -52,4 +56,31 @@ public Response getAssetsInfo(@Context final HttpServletRequest request,
...
+    @NoCache
+    @Consumes(MediaType.MULTIPART_FORM_DATA)
+    @Produces({MediaType.APPLICATION_JSON, "application/javascript"})
+    public Response putAsset(
+            @Context final HttpServletRequest request,
+            @Context final HttpServletResponse response,
+            @BeanParam FileUploadData form
+    ) throws DotSecurityException, DotDataException, IOException {
+
+        final InitDataObject initDataObject = new WebResource.InitBuilder()
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/asset/WebAssetResource.java:73]: nice, some log

**Comment by jdotcms** [comment]
In general we have some security stuff for upload file on the temp api
I am not sure if you should follow the temp api pattern or reimplement the rules here

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/asset/WebAssetHelper.java (hunk 3 above)
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotcms/rest/api/v1/asset/WebAssetHelper.java:374]: done

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/asset/WebAssetResource.java (hunk 4 above)
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotcms/rest/api/v1/asset/WebAssetResource.java:104]: done

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/asset/WebAssetResource.java [hunk 5]
```diff
@@ -49,7 +66,186 @@ public Response getAssetsInfo(@Context final HttpServletRequest request,
...
+     * Delete an asset by path
+     * @param request
+     * @param response
+     * @param form
+     * @return
+     * @throws DotSecurityException
+     * @throws DotDataException
+     * @throws IOException
+     */
+    @Path("/archive")
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/asset/WebAssetResource.java:189]: shouldn't be _archive

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/asset/WebAssetResource.java [hunk 6]
```diff
@@ -49,7 +66,186 @@ public Response getAssetsInfo(@Context final HttpServletRequest request,
...
+    @Path("/")
+    @DELETE
+    @JSONP
+    @NoCache
+    @Produces({MediaType.APPLICATION_JSON, "application/javascript"})
+    public Response deleteAsset(
+            @Context final HttpServletRequest request,
+            @Context final HttpServletResponse response,
+            AssetInfoRequestForm form
+    ) throws DotSecurityException, DotDataException, IOException {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/asset/WebAssetResource.java:163]: this is a delete, does it need a form or just a path param?

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/asset/WebAssetResource.java [hunk 7]
```diff
@@ -49,7 +66,186 @@ public Response getAssetsInfo(@Context final HttpServletRequest request,
...
+                .requiredBackendUser(true)
+                .requiredFrontendUser(false)
+                .requestAndResponse(request, response)
+                .rejectWhenNoUser(true).init();
+
+        final User user = initDataObject.getUser();
+        helper.archiveAsset(form.assetPath(), user);
+        Logger.info(this,
+                String.format("User [%s] deleted asset for path [%s] ", user.getUserId(), form.assetPath()));
+        return Response.ok().build();
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/asset/WebAssetResource.java:210]: you may return here ResponseEntityBooleanView in order to have typed responses for swagger

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/asset/WebAssetResource.java [hunk 8]
```diff
@@ -49,7 +66,186 @@ public Response getAssetsInfo(@Context final HttpServletRequest request,
...
+     * Delete a folder by path
+     * @param request
+     * @param response
+     * @param form
+     * @return
+     * @throws DotSecurityException
+     * @throws DotDataException
+     * @throws IOException
+     */
+    @Path("/folder")
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/asset/WebAssetResource.java:224]: folders

Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/rest/api/v1/asset/WebAssetResource.java:224]: same may be path param instead of a form

**Review by jcastro-dotcms — APPROVED** [review]

File: dotCMS/src/integration-test/java/com/dotcms/rest/api/v1/asset/AssetPathResolverImplIntegrationTest.java [hunk 9]
```diff
@@ -169,4 +168,41 @@ public void Test_Parse_Asset_Under_Root() throws DotDataException, DotSecurityEx
     }
 
 
+    /**
+     * Given scenario: We request a ur with a non-existing folder
```
Comment (nollymar) [inline dotCMS/src/integration-test/java/com/dotcms/rest/api/v1/asset/AssetPathResolverImplIntegrationTest.java:172]: a **url** ?

**Review by nollymar — CHANGES_REQUESTED** [review]
@fabrizzio-dotCMS you should include your new tests to the MainSuite2b as well. The way tests are executed was modified for performance reasons

**Review by jgambarios — APPROVED** [review]

**Review by nollymar — APPROVED** [review]

**Pull request merged**